exec/cnex/array.c

```diff
@@ -9,29 +9,41 @@
 #include "cell.h"
 #include "util.h"
 
+// ToDo: make dictionary and array pre-malloc sizes adjustable with a command line parameter.
+// ToDo: Profile cnex using various starting sizes to figure out what the best starting value would be.
+static const size_t ARRAY_INITIAL_SIZE = 8;  // Reasonable starting value.  Will get us in the ballpark.
+
 Array *array_createArray(void)
 {
     Array *a = malloc(sizeof(Array));
     if (a == NULL) {
         fatal_error("Unable to allocate array.");
     }
 
-    a->size = 0;
     a->data = NULL;
+    a->size = 0;
+    a->max = ARRAY_INITIAL_SIZE;
+    a->data = malloc(a->max * sizeof(Cell));
     a->refcount = 1;
     return a;
 }
 
 Array *array_createArrayFromSize(size_t iElements)
 {
-    Array *a = array_createArray();
+    Array *a = malloc(sizeof(Array));
+    if (a == NULL) {
+        fatal_error("Unable to allocate array for size.");
+    }
 
+    a->data = NULL;
     a->size = iElements;
-    a->data = malloc(sizeof(Cell) * a->size);
-    a->refcount = 1;
+    a->max = (a->size > 0 ? a->size * 2 : ARRAY_INITIAL_SIZE);
```

Comment:
You don't need to preallocate 2x the storage here. You're creating an array with a known number of elements, like maybe a copy of another array, or perhaps a literal constant. Allocate only what is required here, and then later IF you start adding new elements, then preallocate some more.

Reply:
Ok.  That makes sense.  I will keep the check for `> 0` however, since some of the array functions start off with a zero-size array, and start appending elements, so I want to have space already allocated for elements.  But if we're called with fixed size, then correct; we're most likely performing a copy, or a CONSA.  I put this logic in originally for the module init code cnex handles on startup, mainly.